### PR TITLE
Refactor database `DiscoveryResourceChecker`

### PR DESCRIPTION
**Pull request by greedy52**
Moving `discoveryResourceChecker` from `lib/srv/db/watcher.go` to `lib/srv/db/cloud/resource_checker.go`

This changes prepares for adding database URL validation as more checkers will be implemented

### CODE REVIEW

File: lib/srv/db/cloud/resource_checker_credentials.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// credentialsChecker performs some quick checks to see whether this database
+// agent can handle the incoming database wrt to the agent's credentials.
+//
+// Note that this checker warns the user with suggestions on how to configure
+// the credentials correctly instead of returning errors.
+type credentialsChecker struct {
```
Comment (greedy52) [inline lib/srv/db/cloud/resource_checker_credentials.go:41]: moved from watcher.go without logic change

File: lib/srv/db/cloud/resource_checker.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	"github.com/gravitational/teleport/lib/cloud"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// DiscoveryResourceChecker defines an interface for checking database
+// resources created by the discovery service.
+type DiscoveryResourceChecker interface {
+	// Check performs required checks on provided database resource before it
+	// gets registered.
+	Check(ctx context.Context, database types.Database) error
```
Comment (greedy52) [inline lib/srv/db/cloud/resource_checker.go:36]: moved from watcher.go but now returns `error`

**Review by GavinFrazar — APPROVED** [review]
LGTM

as for the data race detected during the test, I believe this is caused by the resource watcher's collector returning all of its collected resources every time a database is created:
https://github.com/gravitational/teleport/blob/3ecc0f01301d8cf221cfc5709c5a9d8f5e31c15c/lib/services/watcher.go#L864-L873

Not sure why the extra test case you added caused the race detector to trip now - I can't reproduce a data race failure if I remove the extra test case. I don't think it's caused by these changes though

**Comment by greedy52** [comment]
> Details

I believe the race happens because the last test case and the new test case race on `db5` that's stored in the test auth server. Let me work on a fix.

**Review by smallinsky — APPROVED** [review]

**Comment by greedy52** [comment]
@GavinFrazar @smallinsky 

I've added a fix for the race [d77047f](https://github.com/gravitational/teleport/pull/29864/commits/d77047fe733e041f21aa4ddb14dee623bf6adbea). 

The idea is to ensure that we don't change the attributes of databases in `s.monitoredDatabases`. And we use `database.Copy()` for `proxiedDatabases`. Attributes of databases in `proxiedDatabases` can be changed but only to "status" not "spec".

Let me know what you think. Thanks!

**Comment by GavinFrazar** [comment]
we should backport the race fix to v13 as well

**Pull request merged**